### PR TITLE
ENH: Allow ArrowDtype(pa.string()) to be compatable with str accessor

### PR DESCRIPTION
**Pull request by mroeschke**
Reboot of #50325

Notable change was that `pandas.core.strings.BaseStringArrayMethods` in the `__init__.py` was causing a circular import when used in `ArrowExtensionArray`, so needed to remove the import into `__init__.py.` This breaks one of our dask tests but it has been fixed downstream https://github.com/dask/dask/pull/9907

### CODE REVIEW

**Comment by jbrockmendel** [comment]
Looks like dask-related doctests are failing.  do we need to pin a version somewhere?

File: pandas/tests/strings/test_strings.py [hunk 1]
```diff
@@ -599,7 +599,7 @@ def test_normalize_index():
     ],
 )
 def test_index_str_accessor_visibility(values, inferred_type, index_or_series):
-    from pandas.core.strings import StringMethods
+    from pandas.core.strings.accessor import StringMethods
```
Comment (jbrockmendel) [inline pandas/tests/strings/test_strings.py:602]: any reason for this not to be at the top of the file?

File: pandas/core/groupby/ops.py [hunk 2]
```diff
@@ -405,7 +406,7 @@ def _reconstruct_ea_result(
         """
         dtype: BaseMaskedDtype | StringDtype
 
-        if isinstance(values.dtype, StringDtype):
+        if is_string_dtype(values.dtype):
```
Comment (jbrockmendel) [inline pandas/core/groupby/ops.py:409]: doesnt make a difference here bc we know we have EA, but we should have a func like is_string_dtype that excludes object

File: pandas/core/groupby/ops.py [hunk 3]
```diff
@@ -96,6 +96,7 @@
 )
 
 if TYPE_CHECKING:
+    from pandas.core.arrays.string_ import StringDtype
```
Comment (jbrockmendel) [inline pandas/core/groupby/ops.py:99]: the annotation that uses this may no longer be accurate?

File: pandas/core/strings/base.py [hunk 4]
```diff
@@ -246,7 +246,9 @@ def _str_removesuffix(self, suffix: str) -> Series:
         pass
 
     @abc.abstractmethod
-    def _str_split(self, pat=None, n=-1, expand: bool = False):
+    def _str_split(
+        self, pat=None, n=-1, expand: bool = False, regex: bool | None = None
```
Comment (jbrockmendel) [inline pandas/core/strings/base.py:250]: is it documented somewhere what regex=None means?

Comment (mroeschke) [reply pandas/core/strings/base.py:250]: Yup, this is in the public docstring https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.str.split.html

File: pandas/core/groupby/ops.py (hunk 3 above)
Comment (mroeschke) [reply pandas/core/groupby/ops.py:100]: Good point. Added to `| ArrowDtype` to the annotation where this is used

File: pandas/tests/strings/test_strings.py (hunk 1 above)
Comment (mroeschke) [reply pandas/tests/strings/test_strings.py]: Don't think so. Moved to the top

**Comment by mroeschke** [comment]
I think the docbuild will fail until dask releases a new version.

@jrbourbeau typically how often does dask release a new version?

**Comment by jrbourbeau** [comment]
> @jrbourbeau typically how often does dask release a new version?

We release every two weeks. In fact, the latest release was pushed out a couple of hours ago

**Comment by mroeschke** [comment]
Oh perfect looks like this PR picked up the dask update and the tests are passing now!

**Comment by mroeschke** [comment]
Any other comments here? (Failures are unrleated)

File: pandas/tests/extension/test_arrow.py [hunk 5]
```diff
@@ -1594,6 +1595,340 @@ def test_searchsorted_with_na_raises(data_for_sorting, as_series):
...
+        ["rfind", ("abc",)],
+        ["split", ()],
+        ["rsplit", ()],
+        ["translate", ("abc",)],
+        ["wrap", ("abc",)],
+    ],
+)
+def test_str_unsupported_methods(method, args):
+    ser = pd.Series(["abc", None], dtype=ArrowDtype(pa.string()))
+    with pytest.raises(NotImplementedError, match=f"str.{method} not supported."):
```
Comment (jbrockmendel) [inline pandas/tests/extension/test_arrow.py:1928]: should the message be string[pyarrow]-specific?

File: pandas/core/arrays/arrow/array.py [hunk 6]
```diff
@@ -1463,6 +1467,285 @@ def _replace_with_mask(
...
+        self, pat: str, case: bool = True, flags: int = 0, na: Scalar | None = None
+    ):
+        if not pat.startswith("^"):
+            pat = f"^{pat}"
+        return self._str_contains(pat, case, flags, na, regex=True)
+
+    def _str_fullmatch(
+        self, pat, case: bool = True, flags: int = 0, na: Scalar | None = None
+    ):
+        if not pat.endswith("$") or pat.endswith("//$"):
```
Comment (jbrockmendel) [inline pandas/core/arrays/arrow/array.py:1559]: i see this is already in the string_arrow code, but why "//$"?

File: pandas/tests/extension/test_arrow.py (hunk 5 above)
Comment (mroeschke) [reply pandas/tests/extension/test_arrow.py]: Sure, I noted it specifically with `pd.ArrowDtype(pa.string())`

File: pandas/core/arrays/arrow/array.py (hunk 6 above)
Comment (mroeschke) [reply pandas/core/arrays/arrow/array.py:1559]: No sure, as you mentioned I just copied it over from the string_arrow code

Comment (jbrockmendel) [reply pandas/core/arrays/arrow/array.py:1559]: OK.  Best guess is it should be backslashes, meant to exclude escaped dollar signs.  OK to consider out of scope.

**Review by jbrockmendel — APPROVED** [review]
LGTM

**Pull request merged**